lib/api/core/hotelClerk.js

```diff
@@ -191,445 +184,406 @@ function HotelClerk (kuzzle) {
    * @param {Request} request
    * @returns {Promise} resolve an object with collection, rooms, subscribers
    */
-  this.listSubscriptions = function hotelClerkListSubscriptions (request) {
-    return new Bluebird(resolve => {
-      const list = {};
-
-      async.each(Object.keys(this.rooms), (roomId, callback) => {
-        const
-          room = this.rooms[roomId],
-          rightRequest = new Request({
-            action: 'search',
-            controller: 'document',
-            index: room.index,
-            collection: room.collection
-          }, request.context);
-
-        return request.context.user.isActionAllowed(rightRequest, kuzzle)
+  listSubscriptions (request) {
+    const
+      list = {},
+      promises = [],
+      dslIndex = this.kuzzle.dsl.storage.filtersIndex;
+
+    for (const index of Object.keys(dslIndex)) {
+      for (const collection of Object.keys(dslIndex[index])) {
+        const isAllowedRequest = new Request({
+          controller: 'document',
+          action: 'search',
+          index,
+          collection
+        }, request.context);
+
+        promises.push(request.context.user.isActionAllowed(isAllowedRequest, this.kuzzle)
           .then(isAllowed => {
             if (!isAllowed) {
-              return callback(null);
+              return;
             }
 
-            if (!list[room.index]) {
-              list[room.index] = {};
+            if (!list[index]) {
+              list[index] = {};
+            }
+            if (!list[index][collection]) {
+              list[index][collection] = {};
             }
 
-            if (!list[room.index][room.collection]) {
-              list[room.index][room.collection] = {};
+            for (const roomId of dslIndex[index][collection]) {
+              const room = this.rooms[roomId];
+              // the room may be currently registering in the dsl and not ready in the hotel clerk
+              // or already deleted in from the hotel clerk and still in the dsl
+              if (room) {
+                list[index][collection][roomId] = room.customers.size;
+              }
             }
+          }));
+      }
+    }
 
-            list[room.index][room.collection][roomId] = room.customers.length;
-            callback(null);
-          });
-      }, () => resolve(list));
-    });
-  };
+    return Bluebird.all(promises)
+      .then(() => list);
+  }
 
   /**
-   * Joins an existing room.
+   * This function will delete a user from this.customers, and
+   * decrement the subscribers count in all rooms where he has subscribed to
+   * Call the cleanUpRooms function to manage empty rooms
```

Comment:
nitpicking: It seems we do not call `cleanupRooms` anymore (the method is deleted)